### PR TITLE
[FLINK-32388]Add the ability to pass parameters to CUSTOM PartitionCommitPolicy

### PR DESCRIPTION
**Pull request by iceiceiceCN**
## What is the purpose of the change

By allowing the passing of parameters, the custom PartitionCommitPolicy becomes more flexible and customizable. This enables user to enhance their custom PartitionCommitPolicy by including additional functionality, such as passing monitoring parameters to track the files associated with each commit.

## Brief change log

- `SINK_PARTITION_COMMIT_POLICY_CLASS_PARAMETERS` parameter added
- To create a custom commit policy through the PartitionCommitPolicyFactory, you have the option to include `SINK_PARTITION_COMMIT_POLICY_CLASS_PARAMETERS` as a parameter.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): (no)
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
- The serializers: (no)
- The runtime per-record code paths (performance sensitive): (no)
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
- The S3 file system connector: (no)

## Documentation

- Does this pull request introduce a new feature?  (yes)
- If yes, how is the feature documented?  (docs)

Key: **sink.partition-commit.policy.class.parameters** 
Default: (none) 
Type: String 
Description: A (semicolon-separated) list of patterns that for custom policy class. Only work in custom commit policy

### CODE REVIEW

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
@iceiceiceCN I think it makes sense first to discuss this on the mailing list, given that it would introduce a new configuration option. Besides that, this PR lacks documentation. Can you briefly discuss this on the Dev mailing list?

**Comment by iceiceiceCN** [comment]
> @iceiceiceCN I think it makes sense first to discuss this on the mailing list, given that it would introduce a new configuration option. Besides that, this PR lacks documentation. Can you briefly discuss this on the Dev mailing list?

**Comment by iceiceiceCN** [comment]
> @iceiceiceCN I think it makes sense first to discuss this on the mailing list, given that it would introduce a new configuration option. Besides that, this PR lacks documentation. Can you briefly discuss this on the Dev mailing list?

Hi MartijnVisser, I've emailed the Dev mailing list to discuss it, thanks for your guidance. : )

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java [hunk 1]
```diff
@@ -63,11 +75,23 @@ public List<PartitionCommitPolicy> createPolicyChain(
                                             successFileName, fsSupplier.get());
                                 case PartitionCommitPolicy.CUSTOM:
                                     try {
-                                        return (PartitionCommitPolicy)
-                                                cl.loadClass(customClass).newInstance();
-                                    } catch (ClassNotFoundException
-                                            | IllegalAccessException
-                                            | InstantiationException e) {
+                                        if (parameters != null && !parameters.isEmpty()) {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:78]: I'm wondering if change it to 
```
parameters != null
```
Can it also get the constructor with empty arguments?
If so, we should consider to change to `parameters != null` to make code more clear.

**Review by luoyuxia — COMMENTED** [review]
@iceiceiceCN Thanks for contribution. I left minor comment. Also, please add test & doc for this new feature.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java (hunk 1 above)
Comment (iceiceiceCN) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:78]: If users input
`'sink.partition-commit.policy.class.parameters'='',`
parameters != null but parameters will empty,
so yes, it maybe get the constructor with empty arguments.

Comment (iceiceiceCN) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:70]: > I'm wondering if change it to
> 
> ```
> parameters != null
> ```
> 
> Can it also get the constructor with empty arguments? If so, we should consider to change to `parameters != null` to make code more clear.

If users input
`'sink.partition-commit.policy.class.parameters'='',`
parameters != null but parameters will empty,
so yes, it maybe get the constructor with empty arguments.

**Review by luoyuxia — COMMENTED** [review]
@iceiceiceCN Thanks for updating. I left some comments.

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/FileSystemCommitterTest.java [hunk 2]
```diff
@@ -258,6 +263,38 @@ void testPartitionPathNotExist() throws Exception {
         assertThat(outputPath.toFile().list()).isEqualTo(new String[0]);
     }
 
+    @Test
+    void testCustomCommitPolicyWithParameters() {
```
[the remark this answers is not in the document]
Comment (iceiceiceCN) [reply flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/FileSystemCommitterTest.java]: > Btw, I don't think we need the test in here if we have a IT.

Yeah, it's a good idea, thanks.
I have removed UT here and add IT in `FileSystemTableSinkTest`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/file/table/FileSystemTableSinkTest.java [hunk 3]
```diff
@@ -149,4 +163,54 @@ private static String buildSinkTableSql(
     private static String buildInsertIntoSql(String sinkTable, String sourceTable) {
         return String.format("INSERT INTO %s SELECT * FROM %s", sinkTable, sourceTable);
     }
+
+    @Test
+    public void testFileSystemTableSinkWithCustomCommitPolicy() throws Exception {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/file/table/FileSystemTableSinkTest.java:168]: Suggestion, I think the test can be simpiled as following:
```
final String outputTable = "outputTable";
        final String customPartitionCommitPolicyClassName = TestCustomCommitPolicy.class.getName();
        final TableEnvironment tEnv =
                TableEnvironment.create(EnvironmentSettings.inStreamingMode());
        String ddl =
                "CREATE TABLE %s ("
                        + "  a INT,"
                        + "  b STRING,"
                        + "  d STRING,"
                        + "  e STRING"
                        + ") PARTITIONED BY (d, e) WITH ("
                        + "'connector'='filesystem',"
                        + "'path'='/tmp',"
                        + "'format'='testcsv',"
                        + "'sink.partition-commit.delay'='0s',"
                        + "'sink.partition-commit.policy.kind'='custom',"
                        + "'sink.partition-commit.policy.class'='%s',"
                        + "'sink.partition-commit.policy.class.parameters'='test1;test2'"
                        + ")";
        ddl = String.format(ddl, outputTable, customPartitionCommitPolicyClassName);
        tEnv.executeSql(ddl);
        tEnv.executeSql(
                        "insert into outputTable select *"
                                + " from (values (1, 'a', '2020-05-03', '3'), "
                                + "(2, 'x', '2020-05-03', '4'))")
                .await();
        Set<String> actualCommittedPaths =
                TestCustomCommitPolicy.getCommittedPartitionPathsAndReset();
        Set<String> expectedCommittedPaths =
                new HashSet<>(
                        Arrays.asList(
                                "test1test2", "/tmp/d=2020-05-03/e=3", "/tmp/d=2020-05-03/e=4"));
        assertEquals(expectedCommittedPaths, actualCommittedPaths);
```

**Review by luoyuxia — COMMENTED** [review]
@iceiceiceCN Thanks for updating. I left minor comments. Should be good in next iteration.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java [hunk 4]
```diff
@@ -63,11 +70,26 @@ public List<PartitionCommitPolicy> createPolicyChain(
                                             successFileName, fsSupplier.get());
                                 case PartitionCommitPolicy.CUSTOM:
                                     try {
-                                        return (PartitionCommitPolicy)
-                                                cl.loadClass(customClass).newInstance();
+                                        if (!CollectionUtil.isNullOrEmpty(parameters)) {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:73]: The ci fail in here. 
```
Jun 29 04:30:25 04:30:25.403 [ERROR]   Architecture Violation [Priority: MEDIUM] - Rule 'Connector production code must depend only on public API when outside of connector packages' was violated (1 times):
Jun 29 04:30:25 Method <org.apache.flink.connector.file.table.PartitionCommitPolicyFactory.createPolicyChain(java.lang.ClassLoader, java.util.function.Supplier)> calls method <org.apache.flink.util.CollectionUtil.isNullOrEmpty(java.util.Collection)> in (PartitionCommitPolicyFactory.java:73)
```
Seems we can't call `CollectionUtil.isNullOrEmpty` in here.

Comment (luoyuxia) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:73]:
```
parameters != null && !parameters.isEmpty()
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/file/table/FileSystemTableSinkTest.java (hunk 3 above)
Comment (iceiceiceCN) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/file/table/FileSystemTableSinkTest.java:168]: More concise! Thanks for your suggesion. 👍

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java (hunk 4 above)
Comment (iceiceiceCN) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java:73]: Ok, and where can I see this ci fail? I hope to learn further : )

Comment (luoyuxia) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java]: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=50618&view=results

Comment (luoyuxia) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionCommitPolicyFactory.java]: You can see the it in `CI report:`

**Review by luoyuxia — APPROVED** [review]
@iceiceiceCN LGTM. Thanks for the pr. 
@MartijnVisser Do you have any other concern? The discussion thread is in https://lists.apache.org/thread/z6zg1l47hxf8ov4p9ysy675cts653dqv

**Comment by MartijnVisser** [comment]
> @MartijnVisser Do you have any other concern? The discussion thread is in

I don't, except that the CI doesn't pass on the current PR code :)

**Comment by luoyuxia** [comment]
@iceiceiceCN The ci fail. Please fix it, you can run  'mvn spotless:apply' to fix  these violations. After that, i'll help merge..

**Comment by iceiceiceCN** [comment]
@flinkbot run azure

**Comment by iceiceiceCN** [comment]
@luoyuxia Hi, I fixed it with the command, and the ci was success. : )

**Review by MartijnVisser — APPROVED** [review]

**Pull request merged**